### ISSUE TITLE
RAC 3.0 support

### ISSUE THREAD

**Issue by MichaelMcGuire** [Moya/Moya#118]
Were you planning on incorporating ReactiveCocoa extensions based on the 3.0 API?  I'm about to create my own extensions and wouldn't mind sending them along if you were interested.

Thanks for the awesome library!

**Comment by ashfurrow** [comment]
:bow: A pull request with the extensions would be awesome! Let me know what I can do to help.

**Comment by ashfurrow** [comment]
I've not looked too extensively into this, and would appreciate guidance or assistance.

Maybe we could add this new functionality to the _existing_ `ReactiveMoyaProvider` class? Or would we want a 3.0-only class? Since the two APIs are interchangeable, it might make sense to have only one class.

**Comment by nanoxd** [comment]
I think it would be best to have it in one class. Perhaps it would be best to have three versions of the API:

1. ReactiveCocoa 2 (RACSignals)
2. ReactiveCocoa 3 Bridge ([RACSignal -> SignalProducer](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/swift-development/ReactiveCocoa/Swift/ObjectiveCBridging.swift))
3. ReactiveCocoa 3 (SignalProducer)

2 would provide an API for those already using `RACSignal` throughout their app and want to switch over piecemeal to RAC 3.

**Comment by sebinsua** [comment]
Has anybody done any work on this yet?

**Comment by rpowelll** [comment]
I'm starting a new project with Moya and RAC 3.0, so it's likely I'll end up writing these extensions as part of them. If I'm happy with how they turn out, I'd be happy to contribute them back.

**Comment by ashfurrow** [comment]
@rpowell that would be lovely!

**Comment by rpowelll** [comment]
@sammoore Yeah I've been working on a small project on and off over the past few weeks. I'll work on getting my extensions polished up and publish the project on the weekend if all goes as planned.

**Comment by justinmakaila** [comment]
This is all taken care of in github.com/Moya/ReactiveMoya. Unfortunately, there's no CocoaPods support yet, only Carthage.

**Comment by zbeckman** [comment]
What's the current/best configuration for use with Swift 2 (Xcode 7) and RAC? It looks like the Moya swift-2.0 branch, w/ RAC swift-2.0 branch, and a carthage pull of ReactiveMoya?

**Comment by ashfurrow** [comment]
@justinmakaila has implemented a solution on another repository (pasted below). We can't incorporate this change into the library yet because it's a breaking change to existing users, and we need to update Moya to the next major version (3.0.0). I'd like to get a Swift 2, Carthage-compatible release out _first_ (see #215) and then incorporate Justin's work, since it looks great :100: 

```swift
/// Subclass of MoyaProvider that returns SignalProducer<MoyaResponse, NSError> instances when requests are made. Much better than using completion closures.
public class ReactiveCocoaMoyaProvider<T where T: MoyaTarget>: MoyaProvider<T> {
    /// Current requests that have not completed or errored yet.
    /// Note: Do not access this directly. It is public only for unit-testing purposes (sigh).
    public var inflightRequests = Dictionary<Endpoint<T>, Signal<MoyaResponse, NSError>>()
    
    /// Initializes a reactive provider.
    override public init(endpointClosure: MoyaEndpointsClosure = MoyaProvider.DefaultEndpointMapping, endpointResolver: MoyaEndpointResolution = MoyaProvider.DefaultEnpointResolution, stubBehavior: MoyaStubbedBehavior = MoyaProvider.NoStubbingBehavior, networkActivityClosure: Moya.NetworkActivityClosure? = nil) {
        super.init(endpointClosure: endpointClosure, endpointResolver: endpointResolver, stubBehavior: stubBehavior, networkActivityClosure: networkActivityClosure)
    }
    
    public func request(token: T) -> SignalProducer<MoyaResponse, NSError> {
        let endpoint = self.endpoint(token)
        
        if let existingSignal = inflightRequests[endpoint] {
            /// returns a new producer which forwards all events of the already existing request signal
            return SignalProducer { sink, disposable in
                /// connect all events of the existing signal to the observer of this signal producer
                existingSignal.observe(sink)
            }
        }
        else {
            /// returns a new producer which starts a new producer which invokes the requests. The created signal of the inner producer is saved for inflight request
            return SignalProducer { [weak self] sink, _ in
                let producer: SignalProducer<MoyaResponse, NSError> = SignalProducer { [weak self] sink, disposable in
                    let cancellableToken = self?.request(token) { data, statusCode, response, error in
                        if let error = error {
                            if let statusCode = statusCode {
                                sendError(sink, NSError(domain: error.domain, code: statusCode, userInfo: error.userInfo))
                            } else {
                                sendError(sink, error)
                            }
                        } else {
                            if let data = data {
                                sendNext(sink, MoyaResponse(statusCode: statusCode!, data: data, response: response))
                            }
                        }
                        sendCompleted(sink)
                    }
                    
                    disposable.addDisposable {
                        if let weakSelf = self {
                            objc_sync_enter(weakSelf)
                            // Clear the inflight request
                            weakSelf.inflightRequests[endpoint] = nil
                            objc_sync_exit(weakSelf)
                            // Cancel the request
                            cancellableToken?.cancel()
                        }
                    }
                }
                
                /// starts the inner signal producer and store the created signal.
                producer |> startWithSignal { [weak self] signal, _ in
                    objc_sync_enter(self)
                    self?.inflightRequests[endpoint] = signal
                    objc_sync_exit(self)
                    /// connect all events of the signal to the observer of this signal producer
                    signal.observe(sink)
                }
            }
        }
    }
    
    public func request(token: T) -> RACSignal {
        return toRACSignal(self.request(token))
    }
}

/// Extension for mapping to a certain response type
public extension ReactiveCocoaMoyaProvider {
    public func requestJSON(token: T) -> SignalProducer<AnyObject, NSError> {
        return request(token) |> mapJSON()
    }
    
    public func requestJSONArray(token: T) -> SignalProducer<NSArray, NSError> {
        return requestJSON(token) |> mapJSONArray()
    }
    
    public func requestJSONDictionary(token: T) -> SignalProducer<NSDictionary, NSError> {
        return requestJSON(token) |> mapJSONDictionary()
    }
    
    public func requestImage(token: T) -> SignalProducer<UIImage, NSError> {
        return request(token) |> mapImage()
    }
    
    public func requestString(token: T) -> SignalProducer<String, NSError> {
        return request(token) |> mapString()
    }
}

/// MoyaResponse free functions

public func filterStatusCode(range: ClosedInterval<Int>) -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<MoyaResponse, NSError>  {
    return { producer in
        return producer |> flatMap(.Latest, { response in
            if range.contains(response.statusCode) {
                return SignalProducer(value: response)
            } else {
                return SignalProducer(error: ReactiveMoyaError.StatusCode(response).toError())
            }
        })
    }
}

public func filterStatusCode(code: Int) -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<MoyaResponse, NSError> {
    return filterStatusCode(code...code)
}

public func filterSuccessfulStatusCodes() -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<MoyaResponse, NSError> {
    return filterStatusCode(200...299)
}

public func filterSuccessfulAndRedirectCodes() -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<MoyaResponse, NSError> {
    return filterStatusCode(200...399)
}

/// Maps the `MoyaResponse` to a `UIImage`
public func mapImage() -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<UIImage, NSError> {
    return { producer in
        return producer |> flatMap(.Latest, { response in
            if let image = UIImage(data: response.data) {
                return SignalProducer(value: image)
            } else {
                return SignalProducer(error: ReactiveMoyaError.ImageMapping(response).toError())
            }
        })
    }
}

/// Maps the `MoyaResponse` to JSON
public func mapJSON() -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<AnyObject, NSError> {
    return { producer in
        return producer |> flatMap(.Latest, { response in
            var error: NSError?
            if let json: AnyObject = NSJSONSerialization.JSONObjectWithData(response.data, options: .AllowFragments, error: &error) {
                return SignalProducer(value: json)
            } else {
                return SignalProducer(error: ReactiveMoyaError.JSONMapping(response).toError())
            }
        })
    }
}

/// Maps a JSON object to an NSArray
public func mapJSONArray() -> SignalProducer<AnyObject, NSError> -> SignalProducer<NSArray, NSError> {
    return { producer in
        return producer |> flatMap(.Latest, { json in
            if let json = json as? NSArray {
                return SignalProducer(value: json)
            } else {
                return SignalProducer(error: ReactiveMoyaError.JSONMapping(json).toError())
            }
        })
    }
}

/// Maps a JSON object to an NSDictionary
public func mapJSONDictionary() -> SignalProducer<AnyObject, NSError> -> SignalProducer<NSDictionary, NSError> {
    return { producer in
        return producer |> flatMap(.Latest, { json in
            if let json = json as? NSDictionary {
                return SignalProducer(value: json)
            } else {
                return SignalProducer(error: ReactiveMoyaError.JSONMapping(json).toError())
            }
        })
    }
}

/// Maps the `MoyaResponse` to a String
public func mapString() -> SignalProducer<MoyaResponse, NSError> -> SignalProducer<String, NSError> {
    return { producer in
        return producer |> flatMap(.Latest, { response in
            if let string =  NSString(data: response.data, encoding: NSUTF8StringEncoding) as? String {
                return SignalProducer(value: string)
            } else {
                return SignalProducer(error: ReactiveMoyaError.StringMapping(response).toError())
            }
        })
    }
}
```

**Comment by ashfurrow** [comment]
Fixed in https://github.com/Moya/Moya/pull/224 – merged into the 3.0.0-develop branch.
